src/storage/checkpoints/checkpoint.cpp

```diff
@@ -0,0 +1,144 @@
+
+
+#include "storage/checkpoints/checkpoint.h"
+#include <common/worker_pool.h>
+#include <fstream>
+namespace terrier::storage {
+
+bool Checkpoint::TakeCheckpoint(const std::string &path, catalog::db_oid_t db) {
+  // get db catalog accessor
+  auto accessor = catalog_->GetAccessor(txn_, db);
+  std::unordered_set<catalog::table_oid_t> table_oids = accessor->GetAllTableOids();
+
+  for (const auto &oid : table_oids) {
+    queue.push_back(oid);
+  }
+
+  // initalize threads
+  auto num_threads = 1u;
+  common::WorkerPool thread_pool_{num_threads, {}};
+  thread_pool_.Startup();
+  auto workload = [&](uint32_t worker_id) {
+    // copy contents of table to disk
+    WriteToDisk(path, accessor, db);
+  };
+  for (auto i = 0u; i < num_threads; i++) {
+    thread_pool_.SubmitTask([i, &workload] { workload(i); });
+  }
+  thread_pool_.WaitUntilAllFinished();
+  if (queue.size() > 0) {
+    // the table oid that failed to be backup
+    return false;
+  }
+  return true;
+}
+
+void Checkpoint::WriteToDisk(const std::string &path, const std::unique_ptr<catalog::CatalogAccessor> &accessor,
+                             catalog::db_oid_t db_oid) {
+  while (queue.size() > 0) {
+    queue_latch.lock();
+    if (queue.size() <= 0) {
+      queue_latch.unlock();
+      return;
+    }
+    catalog::table_oid_t curr_table_oid = queue[0];
+    queue.erase(queue.begin());
+    queue_latch.unlock();
+    common::ManagedPointer<storage::SqlTable> curr_table = accessor->GetTable(curr_table_oid);
+
+    std::string out_file = GenFileName(db_oid, curr_table_oid);
+    std::ofstream f;
+
+    auto curr_data_table = curr_table->table_.data_table_;
+    const BlockLayout &layout = curr_data_table->GetBlockLayout();
+    const std::vector<col_id_t> varlens = layout.Varlens();
+    auto column_ids = layout.AllColumns();
+    size_t col_num = column_ids.size();
+    std::cout << "col num: " << col_num << std::endl;
+
+    curr_data_table->blocks_latch_.Lock();
+    const std::list<RawBlock *> blocks = curr_data_table->blocks_;
+    curr_data_table->blocks_latch_.Unlock();
+    f.open(path + out_file, std::ios::binary);
+
+    // record # of blocks in block list, # of varlen col and # of dict col in the table
+    if (f.is_open()) {
+      unsigned long block_num = blocks.size();
+      unsigned long var_col_num = varlens.size();
+      unsigned long dict_col_num = 0;
+      unsigned long insertion_head_index = std::distance(curr_data_table->blocks_.begin(), curr_data_table->insertion_head_);
+      std::cout << "block num: " << block_num << std::endl;
```

Comment:
Maybe the debug logging statements used in terrier could replace all the std::cout statements?

Reply:
Yes! Our code check tools should have thrown an error when if you have `std::cout`. Did it not catch this for you?